src/backend/distributed/planner/distributed_planner.c

```diff
@@ -336,6 +337,7 @@ ListContainsDistributedTableRTE(List *rangeTableList)
 
 		if (IsCitusTable(rangeTableEntry->relid))
 		{
+			WarnUnsupportedIfForeignDistributedTable(rangeTableEntry->relid);
```

Comment:
I'm really not convinced that calling `WarnUnsupportedIfForeignDistributedTable` inside `ListContainsDistributedTableRTE` is a good approach. 

What about:
```C
	else if (CitusHasBeenLoaded())
	{
		needsDistributedPlanning = ListContainsDistributedTableRTE(rangeTableList, maybeHasForeignTable);
		if (needsDistributedPlanning)
		{
			fastPathRouterQuery = FastPathRouterQuery(parse, &distributionKeyValue);
			
			if (maybeHasForeignTable)
			{
			
                 WarnUnsupportedIfForeignDistributedTable(rangeTableEntry->relid);
			}
		
		}
	}
```

Reply:
We actually need a relationId to pass to `WarnUnsupportedIfForeignDistributedTable`, to verify the relation is a distributed foreign table. But here we don't have a relationId, instead we have a list that might contain a relation that is a distributed foreign table.

Besides, there is one more place that we call `ListContainsDistributedTableRTE`, that has nothing to do with foreign tables, so passing such a parameter doesn't make sense in that place.

We can consider iterating the list once more, to find out if it contains any distributed foreign table; but this option adds some unnecessary complexity. Maybe we can avoid this complexity by doing the iteration only when the variable `DistributedForeignTableWarningPrompted` is `false` ?

Something like:
```c
	else if (CitusHasBeenLoaded())
	{
		needsDistributedPlanning = ListContainsDistributedTableRTE(rangeTableList);
		if (needsDistributedPlanning)
		{
			fastPathRouterQuery = FastPathRouterQuery(parse, &distributionKeyValue);

			WarnIfListHasForeignDistributedTable(rangeTableList);
		}
	}
```
and 
```c
static void
WarnIfListHasForeignDistributedTable(List *rangeTableList)
{
	RangeTblEntry *rangeTableEntry = NULL;
	foreach_ptr(rangeTableEntry, rangeTableList)
	{
		if (DistributedForeignTableWarningPrompted)
		{
			return;
		}

		WarnUnsupportedIfForeignDistributedTable(rangeTableEntry->relid);
	}
}
```

Reply:
> Besides, there is one more place that we call ListContainsDistributedTableRTE, that has nothing to do with foreign tables, so passing such a parameter doesn't make sense in that place.

Similarly, calling `WarnUnsupportedIfForeignDistributedTable` doesn't make sense :)

 Can't we re-iterate the list only when we know there is a foreign table? In majority of the cases `DistributedForeignTableWarningPrompted ` will be false, hence we'd re-iterate unnecessarily with the above suggestion